### ISSUE TITLE
IPv4Address and UNIXAddress not-equal comparison is broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4817]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#4817|
|Type|defect|
|Created|2011-01-22 10:02:46Z|
|Branch|https://github.com/twisted/twisted/tree/address-not-equal-comparison-4817-2|

Both `IPv4Address` and `UNIXAddress` implement `__eq__`, but not `__ne__`.

```
>>> from twisted.internet.address import IPv4Address
>>> IPv4Address('TCP', '127.0.0.1', 10000) == IPv4Address('TCP', '127.0.0.1', 10000) 
True
>>> IPv4Address('TCP', '127.0.0.1', 10000) != IPv4Address('TCP', '127.0.0.1', 10000) 
True
```

Attachments:

* [4817-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afb/afbb340679d1103b326d316ece47049291811c15/61a003416acc8dc922b2c8940ec6cb00b66d556a.patch) (3715 bytes) - added by awclin on 2011-01-22 19:08:55Z - 
* [4817-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afb/afbb340679d1103b326d316ece47049291811c15/5134084bfd239550ea296ff530ba3ef8bc354bd9.patch) (9242 bytes) - added by awclin on 2011-01-24 10:17:35Z - 
* [4817-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afb/afbb340679d1103b326d316ece47049291811c15/895f4fa843a6299e1a2d06fdcc60abe85ccd7441.patch) (5360 bytes) - added by awclin on 2011-01-30 13:20:12Z - 1st patch against address-not-equal-comparison-4817

<details><summary>Searchable metadata</summary>

```
trac-id__4817 4817
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__branches_address_not_equal_comparison_4817_2 branches/address-not-equal-comparison-4817-2
branch_author__exarkun__awclin exarkun, awclin
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1295690566000000 1295690566000000
changetime__1297654550000000 1297654550000000
version__None None
owner__ 
cc__awclin
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Using `twisted.python.util.FancyEqMixin` should make this a snap to fix.

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|awclin commented|
|-|-|

The test returned a failure ('None' result instead of false) prior to the patch, and passed after.  I believe I'm testing the explicit problem detailed in the ticket, though not sure if its advisable to increase the coverage in this area as well?

Post patch, i noticed a test failing in twisted/test/test_udp.py which i've updated to reflect the changes to the equality operator.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [30566]) Branching to 'address-comparison-4817'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [30567]) Apply 4817-1.patch 

refs [#4817](https://github.com/twisted/twisted/issues/4817)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to awclin|
|-|-|

Thanks!

  1. Since `testOldAddress` no longer emits a warning, the suppression for it should be deleted.  Also, can you add a docstring and rename it to `test_oldAddress`?  Any time old code is being changed that doesn't conform to some part of the coding standard is a good time to bring it up to date.
  1. Any modified file should have its copyright date updated as well.
  1. In `test_address.py`, it would probably be good to use `assertTrue` for the equality case, to make it more obvious that both `==` and `!=` are being tested.
  1. It looks like `UNIXAddress` had some untested behavior - the `os.path.samepath` check when the two names were not equal.  We should try to keep this behavior.  Can you add a test exercising it and re-add it?  Hopefully `FancyEqMixin` is still usable, but `UNIXAddress.__eq__` probably needs to add the extra `samefile` check if the inherited `__eq__` returns `False`.
  1. This is probably a good time to delete the deprecated `__getitem__` and `__getslice__` methods on these classes.  The change already removes the deprecated comparison to tuple; removing the whole mess at once would be nice.
  1. Can you also add _two_ news fragments for this change?  One should describe the bug being fixed (comparison of address objects).  The other should mention the removed features - tuple comparison and indexing.  News fragments are described [here](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).

I checked the patch into a branch.  Please make future patches against that branch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|awclin commented|
|-|-|

Thanks for the detailed review, and apologies for missing a couple of things that were documented in the developer guidelines first time around. The new patch addresses all the points you've raised.

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|awclin removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30574]) Branching to 'address-not-equal-comparison-4817'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30575]) Apply 4817-1.patch and 4817-2.patch

refs [#4817](https://github.com/twisted/twisted/issues/4817)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [30576]) Apply the rest of them

refs [#4817](https://github.com/twisted/twisted/issues/4817)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to awclin|
|-|-|

Thanks!  I noticed a couple more things:

  1. `_bwHack` looks like it's totally obsolete now.  I think we can delete the attribute and the related comments.  The `__init__` parameter should stay, I guess, and be deprecated separately: /
  1. `test_differentNamesComparison` looks like something we should be applying to _all_ of the address types.  There could be a `buildDifferentAddress` hook in addition to `buildAddress` and the base class could define a test that asserts that the two addresses are not equal to each other.

I checked the patches into a branch, so please make the next patch against that branch.  Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="awclin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|awclin removed owner|
|-|-|

comments addressed

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [30643]) Branching to 'address-not-equal-comparison-4817-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [30645]) Apply 4817-3.patch 

refs [#4817](https://github.com/twisted/twisted/issues/4817)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks!

  1. Whenever we introduce deprecations, we should make sure that Twisted itself doesn't trigger them (we've failed to do this in the past, and the results are annoyingly visible in the output of `trial twisted`).  So in this case, all of the places in Twisted that instantiate an address should stop passing the final argument.
  1. Test method docstrings don't need to say "Assert ..." or "Ensure ...".  The docstrings should just explain the feature being tested.
  1. The new test methods and helpers should have docstrings as well.
  1. `UNIXAddressTestCase.buildDifferentAddress` should probably return a fixed address, like the IPv4 cases do.
  1. lines should be less than 80 columns

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun removed owner|
|-|-|

I addressed these points in r30646.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

[Build results http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/address-not-equal-comparison-4817-2]

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [30647]) Conditionally define UNIXAddress.__eq__, only if samefile is defined (so, not on Windows).

refs [#4817](https://github.com/twisted/twisted/issues/4817)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set owner to @exarkun|
|-|-|

The `samefile` check seems a bit weird to me; won't this change depending on the presence of symlinks and the state of the real filesystem?  The TCP addresses don't attempt to determine if their IPs are different interfaces on the same host.

But, I'll trust your judgement if you want to leave it as-is; it otherwise looks fine to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

The `samefile` doesn't seem too awesome to me, but since it was there before I want to leave it for now.  We may want to get rid of it soon though.  I don't think it can ever actually be invoked.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [30743]) Merge address-not-equal-comparison-4817-2 

Author: awclin, exarkun
Reviewer: exarkun, glyph
Fixes: [#4817](https://github.com/twisted/twisted/issues/4817)

Remove the deprecated support for indexing `IPv4Address` and `UNIXAddress` instances.
Also deprecate passing the backwards compatibility hack argument to those classes and
stop doing it everywhere in Twisted.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
